Sources/Core/Shared/GroupOperation.swift

```diff
@@ -122,10 +142,53 @@ public class GroupOperation: Operation, OperationQueueDelegate {
      - parameter operations: an array of `NSOperation` instances.
      */
     public func addOperations(additional: [NSOperation]) {
+        _addOperations(additional, addToOperationsArray: true)
+    }
+
+    private func _addOperations(additional: [NSOperation], addToOperationsArray: Bool = true) {
+
         if additional.count > 0 {
-            additional.forEachOperation { $0.log.severity = log.severity }
-            queue.addOperations(additional)
-            operations.appendContentsOf(additional)
+
+            let shouldAddOperations = groupFinishLock.withCriticalScope { () -> Bool in
+                guard !isGroupFinishing else { return false }
+                dispatch_group_enter(isAddingOperationsGroup)
+                return true
+            }
+
+            guard shouldAddOperations else {
+                if !finishingOperation.finished {
+                    assertionFailure("Cannot add new operations to a group after the group has started to finish.")
+                }
+                else {
+                    assertionFailure("Cannot add new operations to a group after the group has completed.")
+                }
+                return
+            }
+
+            var handledCancelled = false
+            if self.cancelled {
+                additional.forEachOperation { $0.cancel() }
+                handledCancelled = true
+            }
+            let logSeverity = self.log.severity
+            additional.forEachOperation { $0.log.severity = logSeverity }
+
+            self.queue.addOperations(additional)
+
+            if addToOperationsArray {
+                self._operations.appendContentsOf(additional)
+            }
+
+            if !handledCancelled && self.cancelled {
+                // It is possible that the cancellation happened before adding the
+                // additional operations to the operations array.
+                // Thus, ensure that all additional operations are cancelled.
+                additional.forEachOperation { if !$0.cancelled { $0.cancel() } }
+            }
+
+            groupFinishLock.withCriticalScope {
+                dispatch_group_leave(isAddingOperationsGroup)
+            }
         }
```

Comment:
Not silly - an earlier commit used a lock (and, thus, closure) around all of that code before switching to finer-grained locks. Tweaking shortly.